### ISSUE TITLE
HttpRequestData should preserve the capitalization of custom http header keys

### ISSUE THREAD

**Issue by dvdgutzmann** [Azure/azure-functions-host#7654]
Hi all!

I have a function running .NET 5 in an isolated process. When the HTTP trigger is hit with a request containing some custom http header key (ie. My-SPECIAL-CaSe-sensitive-Header), the capitalization seems to be lost when accessing the header via `HttpRequestData`. The `BindingContext`, however, holds the custom header key with the original capitalization. `HttpRequestData `should also preserve the original capitalization of custom http headers keys.

![image](https://user-images.githubusercontent.com/20907834/130434827-3d2fa0e7-d24f-4fbf-b58b-bc3f980b40ba.png)

![image](https://user-images.githubusercontent.com/20907834/130434850-194d1e9b-e12a-4562-8e6c-bc092b5720c8.png)

Thank you!
David

**Comment by fabiocav** [comment]
Transferring this to the host as it is something the RPC layer is currently doing and impacts all languages. 

One option here to preserve the behavior for existing languages is to add a language capability to allow the worker to define the behavior and preserve casing.

**Comment by pragnagopa** [comment]
> do any of you have any context on why we lowercase headers in the host?

This was specifically added for Node and is hidden behind a capability flag.  I think commit - https://github.com/Azure/azure-functions-nodejs-worker/commit/544f7e028409d8be54de85d81ba69d4dbeaf5138#diff-90ce154251852a9af8c28a4b0779fcf6c3a031643eadecd3f9437cbb1b5a6877 has more details. Please take a look

**Comment by fabiocav** [comment]
Makes sense, but it was combined with the nullable support, so we may need to be more granular here to preserve casing in those scenarios.

**Comment by fabiocav** [comment]
Work discussed:

- Introduce a capability flag (e.g. `preserveheadercasing`)
- Preserve header casing when the flag is specified by the worker.

**Comment by javafrog** [comment]
Could this also be the cause for the effect that when developing a Java-based Function, I also have to request HTTP headers in lower case?

I find this behavior highly unusual and it has already given us a lot of headaches. For example, when requesting an "Authorization" header, unlike in any other environment, I have to fetch it as "authorization".

**Comment by Geronius** [comment]
Please is there some solution for this lowering the headernames already? This has lots of impact for me moving from core5 inproc.
